---
title: "Debug"
author: "Jennifer Blanc"
date: "2022-10-03"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(tidyverse)
library(pgenlibr)
```

# Debug Pipeline 

## Look at test vector 

```{r}
meta <- fread("../output/transfer/hgdp_wgs.20190516.metadata.txt")
Tvec <- fread("../output/transfer/Tvec_cordinates.txt")
```


## Computing TGWAS 

ALL HGDP samples 
```{r}
TGWAS <- fread("../output/transfer/TGWAS.txt.gz")
plot(TGWAS$longitude, TGWAS$latitude)
```
Seems like $T^{GWAS}$ is capturing structure in the UKBB

## Running the GWAS 

Uncorrected (chr 3)
```{r}
u <- fread("../output/transfer/ukb_imp_chr3_v3.Height.glm.linear.gz")
u$P <- as.numeric(u$P) 
plot(-1 * log10(u$P))
abline(a = -1 * log10(5e-8), b =0, col = "red")
```
Corrected TGWAS-Lat
```{r}
lat <- fread("../output/transfer/ukb_imp_chr3_v3-Lat.Height.glm.linear.gz")
lat$P <- as.numeric(lat$P) 
plot(-1 * log10(lat$P))
abline(a = -1 * log10(5e-8), b =0, col = "red")
```
Corrected TGWAS-Long
```{r}
long <- fread("../output/transfer/ukb_imp_chr3_v3-Long.Height.glm.linear.gz")
long$P <- as.numeric(long$P) 
plot(-1 * log10(long$P))
abline(a = -1 * log10(5e-8), b =0, col = "red")
```
Check correlation
```{r}
plot(u$BETA,  lat$BETA)
abline(0,1,col="red")
cor(u$BETA, lat$BETA)
```

```{r}
plot(u$BETA,  long$BETA)
abline(0,1,col="red")
cor(u$BETA, long$BETA)
```


## Ascertaining SNPs 

```{r}
a_u <- fread("../output/transfer/ukb_imp_chr3_v3.Height.betas.gz")
plot(u$POS,-1 * log10(u$P))
abline(a = -1 * log10(5e-8), b =0, col = "red")
points(a_u$POS,-1 * log10(a_u$P), col = "red",  pch=24)
```
Clearly the ascertainment is not working the way I thought...test code here
```{r}
# Read in block file
ld_blocks = fread("../output/transfer/fourier_ls-all_parsed.bed")

# Read in all betas
betas_u <- fread("../output/transfer/ukb_imp_chr3_v3.Height.glm.linear.gz")
betas_u$P <- as.numeric(betas_u$P)

# Set p-value threshold 
pt <- as.numeric(5e-8)

# Keep snps under threshold
df_u <- betas_u %>%
  filter(P < pt) 

# Assign SNPs to blocks
df_u$block <- NA
block_chr <- ld_blocks %>% filter(chr == as.numeric(betas_u$`#CHROM`[1]))
if (df_u$POS[1] < as.numeric(block_chr[1,2])) {
  i <- tail(which(df_u$POS < as.numeric(block_chr[1,2])),1)+1 
} else {
  i <- 1
}
bp <- df_u$POS[i]
for (block_num in block_chr$block_number) {
  block_num <- as.numeric(as.character(block_num))
  s <- as.numeric(block_chr %>% filter(block_number == block_num) %>% select(start))
  e <- as.numeric(block_chr %>% filter(block_number == block_num) %>% select(stop))
  df_u <- df_u %>% mutate(block = case_when((POS >= s & POS < e) ~ block_num, TRUE ~ as.numeric(as.character(block))))
}

# Pick the minimum p-value per block
a_u <- df_u %>%
  drop_na() %>%
  group_by(block) %>% arrange(P) %>% slice(n=1)
```

```{r}
plot(betas_u$POS,-1 * log10(betas_u$P))
abline(a = -1 * log10(5e-8), b =0, col = "red")
points(a_u$POS,-1 * log10(a_u$P), col = "red",  pch=24)
```
Fixed Ascertainment

## Calculating Qx 

Check the computation of Qx using EUR 

```{r}
read_genos <- function(geno_prefix, betas_id) {

  pvar <- pgenlibr::NewPvar(paste0(geno_prefix, ".pvar"))
  d1 <- pgenlibr::NewPgen(paste0(geno_prefix, ".pgen"))
  var.ids <- betas_id
  var.indx <- rep(0, length(var.ids))
  for (i in 1:length(var.indx)) {
    var.indx[i] <- pgenlibr::GetVariantsById(pvar,var.ids[i])
  }
  X <- ReadList(d1,var.indx, meanimpute=T) # Check this
  colnames(X) <- var.ids

  return(X)
}

# Function to compute PGS
pgs <- function(X, betas) {

  # Mean center genotypes
  X <- apply(X, 2, function(y) y - mean(y))

  # Comput PGS
  Bhat_strat <- betas
  Z_strat <- X %*% Bhat_strat

  # Format output
  out <- cbind(Z_strat)
  colnames(out) <- c("STRAT")

  return(out)
}

# Function to calculate Va
calc_Va <- function(geno_mat, es) {

  # Get allele frequency in test panel
  freq <- colMeans(geno_mat) /2

  # Pull out effect sizes only
  effect_size <- es

  # Compute Va
  Va <- 2 * sum((effect_size)^2 * freq * (1 - freq))
  return(Va)
}

# Function to calculate Qx
calc_Qx <- function(pgs, tvec, Va, lambda_T) {

  # Compute Qx Strat
  Ztest <- t(tvec) %*% pgs
  Qx_strat <- (t(Ztest) %*% Ztest) / (Va*lambda_T)

  return(Qx_strat)
}

# Function to flip effect sizes
flip <- function(betas) {
  new_betas <- sample(c(-1,1), length(betas),  replace = T) * betas
  return(new_betas)
}

# Function to flip effect sizes and recompute Qx
en <- function(betas, tvec, Va, X, lambda_T) {

  # Flip effect sizes
  b <- flip(betas)

  # Calculate PGS
  prs <- pgs(X, b)

  # Calculate Qx
  Qx <- calc_Qx(prs, tvec, Va, lambda_T)

  return(Qx)
}


# Read in betas
betas <- fread("../output/transfer/ukb_imp_all_v3.Height-Long.betas")


# Read in Genotypes
X <- read_genos("/Users/jenniferblanc/Desktop/hgdp_wgs.20190516.full", betas_id = betas$ID)

# Compute PGS
sscore <- pgs(X, betas$BETA_Strat)

# Compute Va
Va <- calc_Va(X, betas$BETA_Strat)

# Load in Test vectors
TV <- fread("../output/transfer/Tvec_cordinates.txt")

# Load in LambdaT
lambdaT <- fread("../output/transfer/Lambda_T.txt")

# Compute Qx
Qx_lat <- calc_Qx(sscore, TV$latitude,  Va, lambdaT$latitude)
Qx_long <- calc_Qx(sscore, TV$longitude,  Va, lambdaT$longitude)

# Generate Empirical null - Lat
num = 1000
redraws <- matrix(0, ncol = 1, nrow = num)
for (i in 1:num){
    redraws[i,] <- en(betas$BETA_Strat, TV$latitude, Va, X, lambdaT$latitude)
}

# Calculate empirical p-values
all_strat <- redraws[,1]
p_strat_en_lat <- length(all_strat[all_strat > Qx_lat[1,1]])/length(all_strat)

# Generate Empirical null - Long
redraws <- matrix(0, ncol = 1, nrow = num)
for (i in 1:num){
  redraws[i,] <- en(betas$BETA_Strat, TV$longitude, Va, X, lambdaT$longitude)
}

# Calculate empirical p-values
all_strat <- redraws[,1]
p_strat_en_long <- length(all_strat[all_strat > Qx_long[1,1]])/length(all_strat)

# Combine results
out <- c(Qx_lat[1,1], Qx_long[1,1], p_strat_en_lat, p_strat_en_long)

out_u <- out
out_lat <- out
out_long <- out
```



