---
title: "Sandbox"
author: "Jennifer Blanc"
date: "2023-09-05"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(tidyverse)
library(ggpubr)
library(latex2exp)
```


## Get non-WBS IDs 

```{r}
WBS <- fread("~/Downloads/WBS.ids")
ALL <- fread("~/Downloads/EthnicBackground_21000.txt")

df <- ALL %>% filter(!IID %in% WBS$IID) %>% filter(EthnicBackground_21000 %in% c(1, 1001, 1002, 1003))
dfOut <- df[,1:2]

fwrite(dfOut,"~/Downloads/nonWBS-White.ids", row.names = F, col.names = T, quote = F, sep = "\t")
```



```{r}
prs <- fread("~/Downloads/FGr-LOCO_sds.all-PC40.sscore")
pheno <- fread("~/Downloads/StandingHeight_Normalized_50.txt")

df <- inner_join(prs, pheno, by = c("#IID"= "IID"))
plot(df$SCORE1_SUM, df$Height)

cor(df$SCORE1_SUM, df$Height)

mod <- lm(df$SCORE1_SUM ~ df$Height)
```









## Look at fastGWA chr22

```{r}
ALL0 <- fread("~/Downloads/French-Sardinian_22-PC0-ALL.fastGWA.gz")
ALL10 <- fread("~/Downloads/French-Sardinian_22-PC10-ALL.fastGWA.gz")
WBS0 <- fread("~/Downloads/French-Sardinian_22-PC0-WBS.fastGWA.gz")
WBS10 <- fread("~/Downloads/French-Sardinian_22-PC10-WBS.fastGWA.gz")

p1 <- ggplot(data = ALL0, aes(x=POS, y=-1 * log10(P))) + geom_point() + ggtitle("ALL - 0")
p2 <- ggplot(data = ALL10, aes(x=POS, y=-1 * log10(P))) + geom_point() + ggtitle("ALL - 10")
p3 <- ggplot(data = WBS0, aes(x=POS, y=-1 * log10(P))) + geom_point() + ggtitle("WBS - 0")
p4 <- ggplot(data = WBS10, aes(x=POS, y=-1 * log10(P))) + geom_point() + ggtitle("WBS - 10")

sv <- ggarrange(p1,p2,p3,p4, nrow =4)
ggsave("~/Desktop/compare.png",sv, width =8, height = 15)
```


## Compare effect sizes

All 0 vs 10
```{r}
plot(ALL0$BETA, ALL10$BETA)
cor(ALL0$BETA, ALL10$BETA)
```

WBS 0 vs 10
```{r}
plot(WBS0$BETA, WBS10$BETA)
cor(WBS0$BETA, WBS10$BETA)
```


ALL 0 vs WBS 0
```{r}
overlap <- intersect(ALL0$SNP, WBS0$SNP)
ALL0_overlap <- ALL0 %>% filter(SNP %in% overlap)
WBS0_overlap <- WBS0 %>% filter(SNP %in% overlap)

plot(ALL0_overlap$BETA, WBS0_overlap$BETA)
cor(ALL0_overlap$BETA, WBS0_overlap$BETA)
```

ALL 10 vs WBS 10 
```{r}
ALL10_overlap <- ALL10 %>% filter(SNP %in% overlap)
WBS10_overlap <- WBS10 %>% filter(SNP %in% overlap)

plot(ALL10_overlap$BETA, WBS10_overlap$BETA)
cor(ALL10_overlap$BETA, WBS10_overlap$BETA)
```





## Compare to Neale lab GWAS

```{r}
dfNL <- fread("~/Downloads/50_raw.gwas.imputed_v3.both_sexes.tsv")
dfNL <- dfNL[13164583:13791467,]

chr22 <- dfNL %>% separate(variant, c("CHR", "POS", "A1", "A2"), ":") %>% filter(CHR == "22")

toPlot <- chr22 %>% filter(pval < 0.05)
toPlot$POS <- as.numeric(toPlot$POS)
ggplot(data = toPlot, aes(x=POS, y=-1 * log10(pval))) + geom_point()
```

## Compare overlapping SNPs 

```{r}
chr22$POS <- as.numeric(chr22$POS) 
NLoverlapp <- intersect(chr22$POS, ALL0_overlap$POS)
NL <- chr22 %>% filter(POS %in% NLoverlapp) %>% group_by(POS) %>% slice(1)
```


ALL 0 vs NL 
```{r}
ALL0_NL <- ALL0 %>% filter(POS %in% NLoverlapp)

plot(ALL0_NL$BETA, NL$beta)
cor(ALL0_NL$BETA,NL$beta)
```

ALL 10 vs NL  
```{r}
ALL10_NL <- ALL10 %>% filter(POS %in% NLoverlapp)

plot(ALL10_NL$BETA, NL$beta)
cor(ALL10_NL$BETA,NL$beta)
```

WBS 0 vs NL 
```{r}
WBS0_NL <- WBS0 %>% filter(POS %in% NLoverlapp)

plot(WBS0_NL$BETA, NL$beta)
cor(WBS0_NL$BETA,NL$beta)
```

WBS 10 vs NL
```{r}
WBS10_NL <- WBS10 %>% filter(POS %in% NLoverlapp)

plot(WBS10_NL$BETA, NL$beta)
cor(WBS10_NL$BETA,NL$beta)
```


## Compare to ALL with 10 PCs and 0.95 threshold

```{r}
df95 <- fread("~/Downloads/fastGWA_0.95.fastGWA")

ggplot(data = df95, aes(x=POS, y=-1 * log10(P))) + geom_point() + ggtitle("ALL - 10 - 0.95")
```
 ALL 0.95 vs 0.05 cutoff 
```{r}
plot(ALL10$BETA, df95$BETA)
cor(ALL10$BETA, df95$BETA)
```

ALL 0.95 vs WBS 0.05
```{r}
df95_overlap <- df95 %>% filter(SNP %in% overlap)

plot(WBS10_overlap$BETA, df95_overlap$BETA)
cor(WBS10_overlap$BETA, df95_overlap$BETA)
```



## Compare Vilma posterior means 

```{r}
df <- fread("~/Downloads/French-Sardinian_22-PC10.estimates.tsv")
df <- df %>% separate(ID, c("ID", "tmp1", "tmp2"), "_")

df2 <- inner_join(WBS10, df, by = c("SNP" = "ID", "A1" = "A1", "A2" = "A2")) %>% mutate(sig = case_when(abs(posterior_ukbb)> 0.0001 ~ P,abs(posterior_ukbb)<= 0.0001~ 1))

ggplot(data = df2, aes(x = BETA, y = posterior_ukbb)) + geom_point()

ggplot(data = df2 , aes(x = POS, y = -1 * log10(P))) + geom_point() + geom_point(aes(x = POS, y = -1 * log10(sig)), col = "red")
```

# PGS in HGDP 

```{r}
df <- fread("~/Downloads/FGr-LOCO_French-Sardinian-PC40.all.sscore")
metadata <- fread("~/Downloads/hgdp_wgs.20190516.metadata.txt")

df <- inner_join(df, metadata, by = c("#IID" = "sample"))
```


```{r}
hist(df$SCORE1_SUM)
plot(df$latitude, df$SCORE1_SUM)
```

```{r}
dfR <- df %>% filter(population == "French" | population == "Sardinian")

ggplot(data = dfR, aes(x = population, y = SCORE1_SUM)) + geom_point() + geom_boxplot(alpha = 0.05) + ylab("Height PGS")

ggplot(data = df, aes(x = population, y = SCORE1_SUM)) + geom_point() + geom_boxplot(alpha = 0.05) + ylab("Height PGS")
```

```{r}
dfControl <- fread("~/Downloads/control_French-Sardinian-PC40.all.sscore")

dfControl <- inner_join(dfControl, metadata, by = c("#IID" = "sample"))

dfRControl <- dfControl %>% filter(population == "French" | population == "Sardinian")

ggplot(data = dfRControl, aes(x = population, y = SCORE1_SUM)) + geom_point() + geom_boxplot(alpha = 0.05) + ylab("Height PGS")

ggplot(data = dfControl, aes(x = population, y = SCORE1_SUM)) + geom_point() + geom_boxplot(alpha = 0.05) + ylab("Height PGS")
```



```{r}
dfR %>% group_by(population) %>% summarise(mean = mean(SCORE1_SUM))

dfRControl %>% group_by(population) %>% summarise(mean = mean(SCORE1_SUM))
```


Fancy SDS 

```{r}
df <- fread("../plots/sds/UK10K/WBS/results.txt")
df$pc <- as.character(df$pc)
df <- df %>% mutate(sig = case_when(p <= 0.05 ~ "*", p > 0.05 ~ ""))
df <- df %>% separate("phenotype", c("phenotype", "tmp1", "tmp2"))
df <- df %>% mutate(pheno = case_when(phenotype == "MeanCorpuscularHaemoglobinConcentration" ~ "MCHC", TRUE ~ phenotype))

# Get bonferonni correction column wise  
df <- df %>% group_by(covar, pc) %>% mutate(p.adjust = p.adjust(p, method ="bonferroni"))
df <- df %>% mutate(sig2 = case_when(p.adjust <= 0.05 ~ "*", p.adjust > 0.05 ~ ""))

df$covar <- as.factor(df$covar)
levels(df$covar) <- c("FALSE", "TRUE")
label_names <- list("FALSE" = TeX("No $\\hat{F}_{Gr}$") , "TRUE" = TeX("$\\hat{F}_{Gr}$") )

hospital_labeller <- function(variable,value){
  return(label_names[value])
}

# Figure 
pl <- ggplot(df, aes(y = pheno, x = pc, fill = q)) + geom_tile() + facet_wrap(~covar, labeller = hospital_labeller) + scale_fill_gradient2(low = "brown4", high = "royalblue4", name = TeX("$\\hat{q}$")) + geom_text(aes(label=sig2), color="black", size=7, fontface="bold")  + ylab("") + xlab("Number of PCs") + theme_bw(base_size = 14) + theme(axis.text = element_text(color = "black")) 

pl

ggsave("~/Desktop/SDSfancy.png",pl)
```


Simple SDS 
```{r}
df <- fread("../plots/sds/UK10K/WBS/results_LR.txt")
df$pc <- as.character(df$pc)
df <- df %>% mutate(sig = case_when(p <= 0.05 ~ "*", p > 0.05 ~ ""))
df <- df %>% separate("phenotype", c("phenotype", "tmp1", "tmp2"))

# Get bonferonni correction column wise  
df <- df %>% group_by(covar, pc) %>% mutate(p.adjust = p.adjust(p))
df <- df %>% mutate(sig2 = case_when(p.adjust <= 0.05 ~ "*", p.adjust > 0.05 ~ ""))

df$covar <- as.factor(df$covar)
levels(df$covar) <- c("FALSE", "TRUE")
label_names <- list("FALSE" = TeX("No $\\hat{F}_{Gr}$") , "TRUE" = TeX("$\\hat{F}_{Gr}$") )

# Figure 
pl1 <- ggplot(df, aes(y = phenotype, x = pc, fill = q)) + geom_tile() + facet_wrap(~covar, labeller = hospital_labeller) + scale_fill_gradient2(low = "brown4", high = "royalblue4", name = TeX("$\\hat{q}$")) + geom_text(aes(label=sig2), color="black", size=7, fontface="bold")  + ylab("") + xlab("Number of PCs") + theme_bw(base_size = 14) + theme(axis.text = element_text(color = "black"), plot.title = element_text(hjust = 0.5)) + ggtitle("Effect Sizes")

pl1

# No Sign 
df <- df %>% group_by(covar, pc) %>% mutate(p.adjust_NoSign = p.adjust(p_NoSign))
df <- df %>% mutate(sigNoSign = case_when(p.adjust_NoSign <= 0.05 ~ "*", p.adjust_NoSign > 0.05 ~ ""))

pl2 <- ggplot(df, aes(y = phenotype, x = pc, fill = q_NoSign)) + geom_tile() + facet_wrap(~covar, labeller = hospital_labeller) + scale_fill_gradient2(low = "brown4", high = "royalblue4", name = TeX("$\\hat{q}$")) + geom_text(aes(label=sigNoSign), color="black", size=7, fontface="bold")  + ylab("") + xlab("Number of PCs") + theme_bw(base_size = 14) + theme(axis.text = element_text(color = "black"), plot.title = element_text(hjust = 0.5)) + ggtitle("No Effect Sizes")

pl <- ggpubr::ggarrange(pl1, pl2, nrow = 1, widths = c(1, 1))


ggsave("~/Desktop/SDSSimpleWBS.png",pl, width = 15, height = 7)

```

























```{r}
dfr2 <- fread("../plots/sds/UK10K/WBS/r2_results.txt") 
dfr2 <- dfr2 %>% separate("phenotype", c("phenotype", "tmp1", "tmp2")) %>% select(r, r2, phenotype, pc)

#library(reshape2)
df_melt <- melt(dfr2, id.vars = c("phenotype", "r", "r2"))

df_melt$value <- as.character(df_melt$value)
pl <- ggplot(data = df_melt, aes(x = r, y = phenotype, fill = value)) + geom_bar(stat='identity', position='dodge') + theme_classic(base_size = 14) + theme(axis.text = element_text(color = "black")) + ylab("") + xlab("Pearson's Correlation (r)") + scale_fill_manual(values = c("plum1", "plum3", "plum4"), name = "PCs")
ggsave("~/Desktop/r2.png",pl, height = 5, width= 7)
```



```{r}
df <- fread("../plots/sds/UK10K/ALL/sds_PCs_FGr.txt")

pl <- ggplot(df, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 14) 
pl
ggsave("~/Desktop/var.png",pl, height = 3, width= 7)


pl2 <- ggplot(df, aes(x = PC, y = rho)) + geom_bar(stat = "identity", fill = "aquamarine4") + geom_errorbar(aes(ymin = lc, ymax = up)) + theme_classic(base_size = 14)+ xlab("PC") + ylab(TeX("$\\rho(PC, \\hat{F}_{Gr})$")) 
pl2
ggsave("~/Desktop/rho.png",pl2, height = 3, width= 7)
```








HGDP Simple Lat Long 
```{r}
df <- fread("../plots/hgdp/ALL/ALL/results_LR_LatLong.txt")
df$pc <- as.character(df$pc)
df <- df %>% mutate(sig = case_when(p <= 0.00227 ~ "*", p > 0.00227 ~ ""))
df <- df %>% separate("phenotype", c("phenotype", "tmp1", "tmp2"))
df <- df %>% separate(contrast, c("contrast"))

# Get bonferonni correction column wise  
df<- df %>% group_by(covar, pc, contrast) %>% mutate(p.adjust = p.adjust(p, method = "bonferroni"))
df <- df %>% mutate(sig2 = case_when(p.adjust <= 0.05 ~ "*", p.adjust > 0.05 ~ ""))

df$covar <- as.factor(df$covar)
levels(df$covar) <- c("FALSE", "TRUE")
label_names <- list("FALSE" = TeX("No $\\hat{F}_{Gr}$") , "TRUE" = TeX("$\\hat{F}_{Gr}$") )
df <- df %>% mutate(pheno = case_when(phenotype == "MeanCorpuscularHaemoglobinConcentration" ~ "MCHC", TRUE ~ phenotype))

# Figure 
df1 <- df %>% filter(pc == 0)
pl1 <- ggplot(df1, aes(y = pheno, x = contrast, fill = q)) + geom_tile() + facet_wrap(~covar, labeller = hospital_labeller)+ scale_fill_gradient2(low = "brown4", high = "royalblue4", name = TeX("$\\hat{q}$")) + geom_text(aes(label=sig2), color="black", size=7, fontface="bold")  + ylab("") + xlab("Test Vector") + theme_bw(base_size = 16) + theme(axis.text = element_text(color = "black"), plot.title = element_text(hjust = 0.5), legend.position = "bottom") + ggtitle("0 PCs")
pl1

df05 <- df %>% filter(pc == 2)
pl05 <- ggplot(df05, aes(y = pheno, x = contrast, fill = q)) + geom_tile() + facet_wrap(~covar, labeller = hospital_labeller)+ scale_fill_gradient2(low = "brown4", high = "royalblue4", name = TeX("$\\hat{q}$")) + geom_text(aes(label=sig2), color="black", size=7, fontface="bold")  + ylab("") + xlab("Test Vector") + theme_bw(base_size = 16) + theme(axis.text = element_text(color = "black"), plot.title = element_text(hjust = 0.5), legend.position = "bottom") + ggtitle("2 PCs")
pl05

df2 <- df %>% filter(pc == 10)
pl2 <- ggplot(df2, aes(y = pheno, x = contrast, fill = q)) + geom_tile() + facet_wrap(~covar, labeller = hospital_labeller)+ scale_fill_gradient2(low = "brown4", high = "royalblue4", name = TeX("$\\hat{q}$")) + geom_text(aes(label=sig2), color="black", size=7, fontface="bold")  + ylab("") + xlab("Test Vector") + theme_bw(base_size = 16) + theme(axis.text = element_text(color = "black"), plot.title = element_text(hjust = 0.5), legend.position = "bottom") + ggtitle("10 PCs")
pl2

pl <- ggpubr::ggarrange(pl1,pl05, pl2, nrow = 1)


ggsave("~/Desktop/HGDPLatLong.png",pl, height = 10, width = 15)

#ggplot(df, aes(y = phenotype, x = contrast, fill = q)) + geom_tile() + facet_grid(vars(covar),vars(pc))+ scale_fill_gradient2(low = "brown4", high = "royalblue4", name = TeX("$\\hat{q}$")) + geom_text(aes(label=sig2), color="black", size=7, fontface="bold")  + ylab("") + xlab("Test Vector") + theme_bw(base_size = 14) + theme(axis.text = element_text(color = "black"), plot.title = element_text(hjust = 0.5))

```

```{r}
df <- fread("~/Downloads/SDS_PC_WBS")
df$PC <- as.integer(df$PC)

#pl <- ggplot(df, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 14) 
pl <- ggplot(df[1:40,], aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold')) + geom_hline(yintercept = 1, color = "red", linetype = "dashed", size = 1.5) 
#+ ggtitle("SDS - WBS")
pl
ggsave("~/Desktop/SDSWBS.png",pl, height = 4, width = 8)


pl2 <- ggplot(df, aes(x = PC, y = rho)) + geom_bar(stat = "identity", fill = "aquamarine4") + geom_errorbar(aes(ymin = lc, ymax = up)) + theme_classic(base_size = 22)+ xlab("PC") + ylab(TeX("$\\rho(PC, \\hat{F}_{Gr})$")) 
pl2
ggsave("~/Desktop/rho.png",pl2, height = 4, width= 8)
```



```{r}
df <- fread("~/stratification-data_analysis/plots/hgdp/EUR/ALL/results_LR_pairwise_EUR.txt")
df$pc <- as.character(df$pc)
df <- df %>% mutate(sig = case_when(p <= 0.05 ~ "*", p > 0.05 ~ ""))
df <- df %>% separate("phenotype", c("phenotype", "tmp1", "tmp2"))
df <- df %>% separate(contrast, c("contrast"), ".all") %>% filter(!contrast %in% c("Lat", "Long"))

# Get bonferonni correction column wise  
df <- df %>% group_by(covar, pc, contrast) %>% mutate(p.adjust = p.adjust(p))
df <- df %>% mutate(sig2 = case_when(p.adjust <= 0.05 ~ "*", p.adjust > 0.05 ~ ""))

df$covar <- as.factor(df$covar)
levels(df$covar) <- c("FALSE", "TRUE")
label_names <- list("FALSE" = TeX("No $\\hat{F}_{Gr}$") , "TRUE" = TeX("$\\hat{F}_{Gr}$") )

# Figure 
pl1 <- ggplot(df, aes(y = contrast, x = phenotype, fill = q)) + geom_tile() + facet_wrap(~covar, labeller = hospital_labeller)+ scale_fill_gradient2(low = "brown4", high = "royalblue4", name = TeX("$\\hat{q}$")) + geom_text(aes(label=sig2), color="black", size=7, fontface="bold")  + ylab("Region Comparison") + xlab("Phenotype") + theme_bw(base_size = 14) + theme(axis.text = element_text(color = "black"), plot.title = element_text(hjust = 0.5),axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 7)) + ggtitle("Effect Sizes")

pl1

# No Sign 
df <- df %>% group_by(covar, pc) %>% mutate(p.adjust_NoSign = p.adjust(p_NoSign))
df <- df %>% mutate(sigNoSign = case_when(p.adjust_NoSign <= 0.05 ~ "*", p.adjust_NoSign > 0.05 ~ ""))

pl2 <- ggplot(df, aes(y = contrast, x = phenotype, fill = q_NoSign)) + geom_tile() + facet_wrap(~covar, labeller = hospital_labeller) + scale_fill_gradient2(low = "brown4", high = "royalblue4", name = TeX("$\\hat{q}$")) + geom_text(aes(label=sigNoSign), color="black", size=7, fontface="bold")  + ylab("Region Comparison") + xlab("Phenotype") + theme_bw(base_size = 14) + theme(axis.text = element_text(color = "black"), plot.title = element_text(hjust = 0.5),axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 7)) + ggtitle("No Effect Sizes")

pl2

pl <- ggpubr::ggarrange(pl1, pl2, nrow = 1, widths = c(1, 1))


ggsave("~/Desktop/HGDP_EUR.png",pl, width = 15, height = 8)

```



```{r}
df <- fread("../data/ukbb/ukbb_pc_snps.txt")
df2 <- df %>% separate(ID, c("CHR", "BP"), ":", remove = FALSE)
df2$CHR <- as.numeric(df2$CHR)

dfOdd <- df2 %>% filter(CHR %in% odd) %>% select("ID", "REF")
dfEven <- df2 %>% filter(CHR %in% even) %>% select("ID", "REF")

fwrite(dfOdd, "../data/ukbb/chr-odd_pc_snps.txt", quote = F, col.names = T, row.names = F, sep = "\t")
fwrite(dfEven, "../data/ukbb/chr-even_pc_snps.txt", quote = F, col.names = T, row.names = F, sep = "\t")
```


```{r}
df <- fread("~/Downloads/all_populations_mles.txt")
df <- df %>% filter(CHROM == "22")
ukbb <- fread("~/Downloads/ukb_imp_chr22_v3.afreq")[1:100000,] 
colnames(ukbb)[1] <- "CHROM"
ukbb <- ukbb %>% separate(ID, c("tmp", "POS")) 
ukbb$POS <- as.numeric(ukbb$POS)

data <- inner_join(df, ukbb)
```



```{r}
# PC data
df <- fread("../plots/sds/r2_PC_FGr.txt")
df$PC <- as.integer(df$PC)
df_eo <- df %>% filter(chrtype_gwas == "chr-even" & chrtype_test == "test-odd") %>% mutate(type = "even-odd")
df_all <- df %>% filter(chrtype_gwas == "chr-all" & chrtype_test == "test-all") %>% mutate(type = "all")
plot_df <- rbind(df_eo, df_all)

# Error data
data <- fread("../plots/sds/block_error.txt")
data$error <- 1 - data$varE


pl <- ggplot(plot_df, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold')) + facet_grid(row = vars(gwas), cols = vars(type)) + geom_hline(data = data, aes(yintercept = error), color = "red", linetype = "dashed", size = 1.5) + ggtitle("SDS")
pl
#ggsave("~/Desktop/sds.png", pl, height = 7,  width = 10)

tmp <- plot_df %>% filter(gwas == "WBS" & type == "even-odd")
tmp2 <- data %>% filter(gwas == "WBS")
pl <- ggplot(tmp, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold'))  + geom_hline(data = tmp2, aes(yintercept = error), color = "red", linetype = "dashed", size = 1.5) + ggtitle("SDS") 
ggsave("~/Desktop/sds_tmp.png", pl, height = 5, width = 7)


pl2 <- ggplot(tmp, aes(x = PC, y = rho)) + geom_bar(stat = "identity", fill = "aquamarine4") + geom_errorbar(aes(ymin = lc, ymax = up)) + theme_classic(base_size = 22)+ xlab("PC") + ylab(TeX("$\\rho(PC, \\hat{F}_{Gr})$")) 
ggsave("~/Desktop/sds_tmp.png", pl2, height = 4, width = 8.5)
```


```{r}
# PC data
df <- fread("../plots/adna/r2_PC_FGr.txt")
df$PC <- as.integer(df$PC)
df_eo <- df %>% filter(chrtype_gwas == "chr-even" & chrtype_test == "test-odd") %>% mutate(type = "even-odd")
df_all <- df %>% filter(chrtype_gwas == "chr-all" & chrtype_test == "test-all") %>% mutate(type = "all")
plot_df <- rbind(df_eo, df_all)

# Error data
data <- fread("../plots/adna/block_error.txt")
data$error <- 1 - data$varE


# European Neolithic 
plot_EN <- plot_df %>% filter(contrast == "EUROPEAN_NEOLITHIC")
data_EN <- data %>% filter(contrast == "EUROPEAN")
pl_EN <- ggplot(plot_EN, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold')) + facet_grid(row = vars(gwas), cols = vars(type)) + geom_hline(data = data_EN, aes(yintercept = error), color = "red", linetype = "dashed", size = 1.5) + ggtitle("European Neolithic")
pl_EN
ggsave("~/Desktop/en.png", pl_EN, height = 7, width = 10)

# Bronze Age 
plot_BA <- plot_df %>% filter(contrast == "BRONZE_AGE")
data_BA <- data %>% filter(contrast == "BRONZE")
pl_BA <- ggplot(plot_BA, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold')) + facet_grid(row = vars(gwas), cols = vars(type)) + geom_hline(data = data_BA, aes(yintercept = error), color = "red", linetype = "dashed", size = 1.5) + ggtitle("Bronze Age")
pl_BA
ggsave("~/Desktop/ba.png", pl_BA, height = 7, width = 10)

# Historical 
plot_H <- plot_df %>% filter(contrast == "HISTORICAL")
data_H <- data %>% filter(contrast == "HISTORICAL")
pl_H <- ggplot(plot_H, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold')) + facet_grid(row = vars(gwas), cols = vars(type)) + geom_hline(data = data_H, aes(yintercept = error), color = "red", linetype = "dashed", size = 1.5) + ggtitle("Historical")
pl_H
ggsave("~/Desktop/h.png", pl_H,  height = 7, width = 10)


tmp <- plot_EN %>% filter(gwas == "ALL" & type == "even-odd")
tmp2 <- data_EN %>% filter(gwas == "ALL")
pl <- ggplot(tmp, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold'))  + geom_hline(data = tmp2, aes(yintercept = error), color = "red", linetype = "dashed", size = 1.5) + ggtitle("European Neolithic") 
pl
ggsave("~/Desktop/EN_tmp.png", pl, height = 5, width = 7)


pl2 <- ggplot(tmp, aes(x = PC, y = rho)) + geom_bar(stat = "identity", fill = "aquamarine4") + geom_errorbar(aes(ymin = lc, ymax = up)) + theme_classic(base_size = 22)+ xlab("PC") + ylab(TeX("$\\rho(PC, \\hat{F}_{Gr})$")) 
pl2

ggsave("~/Desktop/EN_tmp.png", pl2, height = 4, width = 8.5)
```


```{r}
# PC data
df <- fread("../plots/hgdp/r2_PC_FGr.txt")
df$PC <- as.integer(df$PC)
df_eo <- df %>% filter(chrtype_gwas == "chr-even" & chrtype_test == "test-odd") %>% mutate(type = "even-odd")
df_all <- df %>% filter(chrtype_gwas == "chr-all" & chrtype_test == "test-all") %>% mutate(type = "all")
plot_df <- rbind(df_eo, df_all)

# Error data
data <- fread("../plots/hgdp/block_error.txt")
data$error <- 1 - data$varE


# Latitude 
plot_Lat <- plot_df %>% filter(contrast == "Lat")
data_Lat <- data %>% filter(contrast == "Lat")
pl_Lat <- ggplot(plot_Lat, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold')) + facet_grid(row = vars(gwas), cols = vars(type)) + geom_hline(data = data_Lat, aes(yintercept = error), color = "red", linetype = "dashed", size = 1.5) + ggtitle("Latitude")
pl_Lat
#ggsave("~/Desktop/Lat.png", pl_Lat,  height = 7, width = 10)

# Longitude 
plot_Long <- plot_df %>% filter(contrast == "Long")
data_Long <- data %>% filter(contrast == "Long")
pl_Long <- ggplot(plot_Long, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold')) + facet_grid(row = vars(gwas), cols = vars(type)) + geom_hline(data = data_Long, aes(yintercept = error), color = "red", linetype = "dashed", size = 1.5) + ggtitle("Longitude")
pl_Long
#ggsave("~/Desktop/Long.png", pl_Long,  height = 7, width = 10)


tmp <- plot_Long %>% filter(gwas == "ALL" & type == "even-odd")
tmp2 <- data_Long %>% filter(gwas == "WBS")
pl <- ggplot(tmp, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold'))  + geom_hline(data = tmp2, aes(yintercept = error), color = "red", linetype = "dashed", size = 1.5) + ggtitle("Longitude") 
ggsave("~/Desktop/long_tmp.png", pl, height = 5, width = 7)


pl2 <- ggplot(tmp, aes(x = PC, y = rho)) + geom_bar(stat = "identity", fill = "aquamarine4") + geom_errorbar(aes(ymin = lc, ymax = up)) + theme_classic(base_size = 22)+ xlab("PC") + ylab(TeX("$\\rho(PC, \\hat{F}_{Gr})$")) 
ggsave("~/Desktop/long_tmp.png", pl2, height = 5.5, width = 8.5)
```


```{r}
# PC data
df <- fread("../plots/hgdp/r2_PC_FGr.txt")
df$PC <- as.integer(df$PC)
plot_df <- df %>% filter(chrtype_gwas == "chr-even" & chrtype_test == "test-odd") %>% mutate(type = "even-odd")

# Error data
data <- fread("../plots/hgdp/block_error.txt")
data$error <- 1 - data$varE
plot_data <- data

#ggplot(plot_df, aes(x = PC, y = r2)) + geom_bar(stat = "identity", fill = "aquamarine3") + ylab(TeX("Variance in $\\hat{F}_{Gr}$ Explained")) + xlab("Cumulative PCs") + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold')) + facet_grid(cols = vars(gwas), rows = vars(contrast)) + geom_hline(data = data, aes(yintercept = error), color = "red", linetype = "dashed", size = 1.5) 

pl <- ggplot(plot_data, aes(x = contrast, y = varE)) + geom_bar(stat = "identity", fill = "purple4") + facet_wrap(~gwas) + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold'), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 8)) + ylab("error")  
pl

tmp <- plot_df %>% filter(PC == 40 )
tmp2 <- inner_join(tmp, plot_data)

pl <- ggplot(tmp2, aes(x = contrast, y = error)) + geom_bar(stat = "identity") + facet_wrap(~gwas) + theme_classic(base_size = 20) + ylim(0,1) + theme( plot.title=element_text( hjust=0.5, face='bold'), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 8)) + ylab("1 - error") + geom_bar(aes(x = contrast,y = (r2*error)), alpha = 0.5, stat = "identity", fill = "red")

ggsave("~/Desktop/continental_error.png", pl,  height = 7, width = 10)
```


```{r}
L <- 300
M <- 200

# Simulate contrasts
r <- runif(L)
r <- scale(r)

# Simulate genotypes
G <- matrix(NA, nrow = M, ncol = L)
for (i in 1:L) {
  G[,i] <- rnorm(M)
}
G <- scale(G, scale = FALSE)
Gnorm <- scale(G, scale = TRUE)

# Do PCA vis SVD
mySVD <- svd(Gnorm)
vecs <- mySVD$u
vals <- mySVD$d
loadings <- mySVD$v

# Calculate FGr
FGr <- Gnorm %*% r *(1/L)

# Calculate D
D <- t(FGr) %*% FGr *(L^2)
D2 <- (t(r) %*% (t(Gnorm) %*% Gnorm) %*% r) 

# Calculate D via loadings
f <- t(r) %*% loadings
fp <- f * vals  
D3 <- sum(fp^2)


# Calculate expD
K <- (Gnorm %*% t(Gnorm)) * (1/(L))
R <- (t(Gnorm) %*% Gnorm) * (1/(M-1))

sum(diag(R)) * (M-1)
sum(diag(K)) * (L)

# Check variance of loadinds 
mean(rowMeans(loadings^2))
1/L
```



```{r}
sim <- function(M, L) {

  # Simulate contrasts
  r <- runif(L)
  r <- scale(r)

  # Simulate genotypes
  G <- matrix(NA, nrow = M, ncol = L)
  for (i in 1:L) {
    G[,i] <- rnorm(M)
  }
  G <- scale(G, scale = FALSE)
  Gnorm <- scale(G, scale = TRUE)
  
  # Calculate FGr
  FGr <- Gnorm %*% r *(1/L)
  
  # Calculate D
  D <- t(FGr) %*% FGr *(L^2)
  
  return(D)
}
```

```{r}
numRep <- 10000
L <- 300
M <- 200

allDs <- rep(NA, numRep)
for (i in 1:numRep) {
  allDs[i] <- sim(M, L)
}
```

```{r}
expD <- (M-1) * L
meanD <- mean(allDs)

hist(allDs)
abline(v = meanD, col = "blue", lwd = 2)
abline(v = expD, col = "red", lwd = 2)
```
```{r}
# Read GWAS Matrix 
library(pgenlibr)
pvar <- NewPvar("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.pvar")
d1 <- NewPgen("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.pgen")
G <- ReadList(d1,seq(1,10905), meanimpute=F)
G <- G[c(1:100, 1901:2000),sample(1:10905, 500)]
Gnorm <- scale(G)
L <- ncol(G)
M <- nrow(G)
r <- runif(L, 0, 1)
r <- scale(r)

# Do PCA vis SVD
mySVD <- svd(Gnorm)
vecs <- mySVD$u
vals <- mySVD$d
loadings <- mySVD$v


# Calculate FGr
FGr <- Gnorm %*% r *(1/L)

# Calculate D
D <- t(FGr) %*% FGr *(L^2)
D2 <- (t(r) %*% (t(Gnorm) %*% Gnorm) %*% r) 

# Calculate D via loadings
f <- t(r) %*% loadings
fp <- f * vals  
D3 <- sum(fp^2)


# Calculate expD
K <- (Gnorm %*% t(Gnorm)) * (1/(L))
R <- (t(Gnorm) %*% Gnorm)
* (1/(M-1))

sum(diag(R)) * (M-1)
sum(diag(K)) * (L)

sum(vals^2)

myE <- eigen(R)
sum(myE$values)

# Check variance of loadinds 
mean(rowMeans(loadings^2))
1/L
```

```{r}
# Read GWAS Matrix 
library(pgenlibr)
pvar <- NewPvar("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.pvar")
d1 <- NewPgen("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.pgen")
G <- ReadList(d1,seq(1,10905), meanimpute=F)
G <- G[c(1:100, 1901:2000),sample(1:10905, 500)]
Gnorm <- scale(G)
L <- ncol(G)
M <- nrow(G)

# Simuate R
r <- runif(L, 0, 1)
r <- r - mean(r)
r <- r / sqrt(sum(r^2))

# Do PCA vis SVD
mySVD <- svd(Gnorm)
vecs <- mySVD$u
vals <- mySVD$d
v <- mySVD$v

# Calculate FGr
FGr <- Gnorm %*% r *(1/L)

# Calculate D
D <-  t(FGr) %*% FGr *(L^2)

# Calculate Expected value of D
expD <- M-1
sum(vals^2) * (1/L)

# Variance explained 
D / ((M-1) * L) 


```



