---
title: "SplitWhiteTestGWAS"
author: "Jennifer Blanc"
date: "2023-11-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Get non-WBS IDs 

```{r}
WBS <- fread("~/Downloads/WBS.ids")
ALL <- fread("~/Downloads/EthnicBackground_21000.txt")

df <- ALL %>% filter(!IID %in% WBS$IID) %>% filter(EthnicBackground_21000 %in% c(1, 1001, 1002, 1003))
dfOut <- df[,1:2]

#fwrite(dfOut,"~/Downloads/nonWBS-White.ids", row.names = F, col.names = T, quote = F, sep = "\t")
```

## Get GWAS and Test panels 

```{r}
# People genotyped in "White" subset
geno <- fread("../data/phenotypes/ukb_imp_chr22_v3.psam")

# People phenotyped for birth coordinate
East <- fread("../data/phenotypes/PlaceOfBirthEastCord_130.txt")
North <- fread("../data/phenotypes/PlaceOfBirthNorthCord_129.txt")

# IDs for white and WBS
ids <- fread("../data/ukbb/White.ids")
wbs <- fread("../data/ukbb/WBS.ids")

# get geno and phenotyped test panel
dfWBS <- inner_join(wbs, East)
dfWBS <- inner_join(dfWBS, North)
dfWBS <- inner_join(dfWBS, geno)

# get geno and phenotyped gwas panel 
dfGWAS <- subset(ids, !(ids$IID %in% wbs$IID))
dfGWAS <- dfGWAS[-1,]
geno$`#FID` <- as.character(geno$`#FID`)
geno$IID <- as.character(geno$IID)
dfGWAS <- inner_join(dfGWAS, geno)
```

## WBS Test panel (tp)

```{r}
set.seed(1212)

# Sample random subsets of test panel size
n <- c(100, 1000, 5000, 50000, 100000)

for (i in 1:length(n)) {
  
  df <- dfWBS %>% sample_n(n[i]) %>% select("#FID", "IID")
  
  name <- paste0("../data/ukbb/tp/t", format(n[i], scientific=F), ".txt")
  fwrite(df, name ,row.names=F,quote=F,sep="\t", col.names = T)
  
}
```

## EUR GWAS panel (gp)

```{r}
set.seed(1313)

# Sample random subsets of test panel size
n <- c(5000, 50000)

for (i in 1:length(n)) {
  
  df <- dfGWAS %>% sample_n(n[i]) %>% select("#FID", "IID")
  
  name <- paste0("../data/ukbb/gp/g", n[i], ".txt")
  fwrite(df, name ,row.names=F,quote=F,sep="\t", col.names = T)
  
}
```


## WBS GWAS panel (wbs)

```{r}
set.seed(1414)

# Sample random subsets of test panel size
n <- c(50000, 100000, 300000)

for (i in 1:length(n)) {
  
  df <- dfWBS %>% sample_n(n[i]) %>% select("#FID", "IID")
  
  name <- paste0("../data/ukbb/gp/wbs", format(n[i], scientific=F), ".txt")
  fwrite(df, name ,row.names=F,quote=F,sep="\t", col.names = T)
  
}
```


## EUR Test panel (eur)

```{r}
set.seed(1515)

# Sample random subsets of test panel size
n <- c(50000)

for (i in 1:length(n)) {
  
  df <- dfGWAS %>% sample_n(n[i]) %>% select("#FID", "IID")
  
  name <- paste0("../data/ukbb/tp/eur", n[i], ".txt")
  fwrite(df, name ,row.names=F,quote=F,sep="\t", col.names = T)
  
}
```


## Splitting WBS 

```{r}
set.seed(1616)

height <- fread("../data/phenotypes/StandingHeight_Normalized_50.txt")
dfWBS <- inner_join(dfWBS, height)

df_test <- dfWBS %>% sample_n(50000) %>% select("#FID", "IID")
name <- paste0("../data/ukbb/tp/sim", format(50000, scientific=F), ".txt")
fwrite(df_test, name ,row.names=F,quote=F,sep="\t", col.names = T)

df_GWAS <- dfWBS %>% filter(!IID %in% df_test$IID) %>% sample_n(300000) %>% select("#FID", "IID")
name <- paste0("../data/ukbb/gp/sim", format(300000, scientific=F), ".txt")
fwrite(df_GWAS, name ,row.names=F,quote=F,sep="\t", col.names = T)
```

```{r}
mod <- lm(df_GWAS$Height ~ df_GWAS$SEX)
df_GWAS$HeightResidual <- mod$residuals

ggplot(df_GWAS, aes(y = PlaceOfBirthNorthCord_129, x = PlaceOfBirthEastCord_130, color = HeightResidual)) + geom_point()

mod2 <- lm(df_GWAS$HeightResidual ~ df_GWAS$PlaceOfBirthNorthCord_129)
summary(mod2)
```

```{r}
ggplot(df, aes(x = PlaceOfBirthNorthCord_129, y = pheno)) + geom_point()
```

```{r}
df <- fread("../plots/Sims/sim50000/sim300000/results_sharp.txt")
```


```{r}
SNPs <- fread("~/Downloads/ukb_snp_qc.txt")


df <- SNPs %>% select("chromosome", "position", "allele1_ref", "allele2_alt", "in_PCA") %>% filter(in_PCA == 1)
df <- df %>% select("chromosome", "position", "allele1_ref")

df$ID <- paste(df$chromosome, df$position, sep=":")
df <- df %>% select("ID",  "allele1_ref")
colnames(df) <- c("ID", "REF")

fwrite(df,"~/stratification-data_analysis/output/transfer/ukbb_pc_snps.txt", row.names = F, col.names = T, quote = F, sep = "\t")
```





```{r}
df <- fread("~/Downloads/qualifications_6138.txt")
dfAge <- fread("~/Downloads/age-finished-education_845.txt")

# Drop first column 
df <- df[,-1]
dfAge <- dfAge[,-1]

# Get max years of education 
maxEdu <- apply(df[, 2:ncol(df)], 1, function(x) ifelse(all(is.na(x)), NA, max(x, na.rm = TRUE)))
maxAge <- apply(dfAge[, 2:ncol(dfAge)], 1, function(x) ifelse(all(is.na(x)), NA, max(x, na.rm = TRUE)))

# Make dataframe
dfEdu <- cbind(df[,1],df[,1], maxEdu, maxAge)
colnames(dfEdu)[1] <- "#FID"
colnames(dfEdu)[2] <- "IID"

# Get rid of codes we don't want 
dfEdu <- dfEdu %>% filter(maxEdu %in% c(1,2,3,4,5,6, -7)) 

# Convert to years 
dfEdu <-dfEdu %>% mutate(years = case_when(maxEdu == 1 ~ 20,
                                           maxEdu == 2 ~ 13,
                                           maxEdu == 3 ~ 10,
                                           maxEdu == 4 ~ 10,
                                           maxEdu == 5 ~ (maxAge-5),
                                           maxEdu == 6 ~ 15,
                                           maxEdu == -7 ~ 7))

```














